### ISSUE TITLE
Font Enumeration API

### ISSUE THREAD

**Issue by chasephillips** [w3ctag/design-reviews#399]
こんにちはTAG!

I'm requesting a TAG review of:

  - Name: Font Enumeration API
  - Specification URL: https://wicg.github.io/local-font-access/
  - Explainer (containing user needs and example code)¹: https://github.com/inexorabletash/font-enumeration
  - GitHub issues (if you prefer feedback filed there): https://github.com/WICG/local-font-access/issues
  - Tests: TBD
  - Primary contacts (and their relationship to the specification): @chasephillips, @inexorabletash

Further details:

  - Relevant time constraints or deadlines: We'd like to discuss this at TPAC (Sep 16, 2019) so getting TAG review before then would be great.  Overall, we plan to Dev Trial in M79 (mid Oct, 2019) and [Origin Trial](https://github.com/GoogleChrome/OriginTrials) in M80 (early Dec, 2019).
  - [x] I have read and filled out the [Self-Review Questionnare on Security and Privacy](https://www.w3.org/TR/security-privacy-questionnaire/). The [assessment is here](https://github.com/inexorabletash/font-enumeration/blob/master/security-privacy.md).
  - [x] I have reviewed the TAG's [API Design Principles](https://w3ctag.github.io/design-principles/)
  - The group where the work on this specification is: TBD

We recommend the explainer to be in [Markdown](https://github.github.com/gfm/). On top of the [usual information expected in the explainer](https://w3ctag.github.io/explainers), it is strongly recommended to add:

  - Links to major pieces of multi-stakeholder review or discussion of this specification: 
  - Links to major unresolved issues or opposition with this specification:

We'd prefer the TAG provide feedback as (please select one):

  - [x] open issues in our GitHub repo for each point of feedback
  - [ ] open a single issue in our GitHub repo for the entire review
  - [ ] leave review feedback as a comment in this issue and @-notify [github usernames]

--------------------------

_Please preview the issue and check that the links work before submitting._ In particular, if anything links to a URL which requires authentication (e.g. Google document), please make sure anyone with the link can access the document.

¹ For background, see our [explanation of how to write a good explainer](https://w3ctag.github.io/explainers).

**Comment by dbaron** [comment]
@ylafon and I are looking at this during a breakout at the Tokyo TAG face-to-face meeting.

One issue that doesn't appear to be addressed in the Security and Privacy questionnaire response is the question of fingerprinting.  Having an API for font enumeration seems like it makes font-based active fingerprinting substantially easier and faster.  That's something that needs to be traded off against the use cases for the feature -- use cases that aren't especially clear in the explainer.

It would probably be helpful if the explainer gave some compelling examples of user-facing features that would be enabled by this web feature.  This would also help understand the question of whether there are other less-powerful APIs that could address the same use cases, for example, APIs designed for finding similar fonts.

It's also worth noting that past studies of font-based fingerprinting had reported entropy that included order data from plugins.  One of the major plugins (I don't remember if it was Java or Flash) had a font enumeration API that returned the fonts in a system-specific sort order that I think was a function of the time the font was installed on the system.  This provided a large amount of additional entropy that was totally unnecessary for the feature.  At the very least it seems important for the spec to prescribe a sort order for the enumeration to avoid a repeat of this problem.

I've still been hoping to find the time to take a somewhat closer look at this, but haven't managed to do that yet.

**Comment by dbaron** [comment]
Also, @ylafon points out that the explainer in #400 seems to provide a rather bad example of the use of this API -- it enumerates all of the fonts on the system just to find "Consolas" -- it seems like that would be much better addressed by an API that lets you ask for "Consolas".

**Comment by hober** [comment]
Labeling as a Fugu-related request since it appears in Fugu's [full list of capabilities](https://bugs.chromium.org/p/chromium/issues/list?can=2&q=proj-fugu&sort=pri&colspec=ID+Pri+M+Stars+ReleaseBlock+Component+Status+Owner+Summary+OS+Modified&x=m&y=releaseblock&cells=ids).

**Comment by hober** [comment]
@plinss and I took another look at this in our Cupertino F2F. We share @dbaron’s & @ylafon’s desire for a “ less-powerful APIs that could address [some of] the same use cases”—in particular, a picker-style API could allow sites to receive access to specific fonts that they need (like Consolas in the previous example) without exposing the list of installed fonts.

@plinss remembers a TPAC conversation from Fukuoka, perhaps in the Houdini TF, where a picker-style API got significant support.

**Comment by atanassov** [comment]
@plinss and I had another look at this topic and its relevance to [Font Table Access API](https://github.com/w3ctag/design-reviews/issues/400). 

There has been no progress we observed and given how tightly coupled the two API proposals are our proposal is to close the issue.

**Comment by inexorabletash** [comment]
re: lack of progress - we deferred work for a while, but are planning to revisit in the very near future.

re: two proposals - we are planning to re-merge the proposals; at one point we thought splitting them made sense to unblock some discussion. Closing one of these out is fine.

**Comment by hober** [comment]
We (@dbaron @hober @plinss) looked at this again during the TAG F2F this week. We've tried to file issues in the WICG/local-font-access repo which capture each of the concerns that have come up during the several TAG breakouts on these two reviews. Given the lack of engagement in these reviews thus far, we don't see any value in keeping these issues open. Please feel free to ping us to re-open or file a new issue when there are significant changes.

* [Non-OpenType fonts #19](https://github.com/WICG/local-font-access/issues/19)
* [Anne's feedback re: Non-goals and alternative approaches #20](https://github.com/WICG/local-font-access/issues/20)
* [Fonts need to be sorted #23](https://github.com/WICG/local-font-access/issues/23)
* [mention additional fingerprinting surface in details of font versions #25](https://github.com/WICG/local-font-access/issues/25)
* [Please add a Table of Contents to the explainer #26]()
* [explainer suggests iterating all the fonts in order to find Consolas #27](https://github.com/WICG/local-font-access/issues/27)

**Comment by oyiptong** [comment]
We've made some progress on the Local Font Access API.

It now combines the Enumeration API and a Font Data Access API.
The repos are all the same, but the explainer is now at https://github.com/wicg/local-font-access

Should this be a new issue, or can I revive this one?

**Comment by plinss** [comment]
@oyiptong Can you give a summary of the changes made and how they address the previous feedback we've given?

**Comment by oyiptong** [comment]
Hi folks, apologies about the delays.

Here's a quick summary:

- The enumeration API now provides an object, `FontMetadata` that is not related to Web Fonts
- Font table data aren't accessible individually anymore
- There's a method to obtain byte-wise font data, i.e. `blob()`
- Use cases are now more elaborately fleshed out

**Comment by hober** [comment]
I wrote [in December 2019](https://github.com/w3ctag/design-reviews/issues/399#issuecomment-561387155):

> @plinss and I[...] share @dbaron’s & @ylafon’s desire for a "less-powerful APIs that could address [some of] the same use cases"—in particular, a picker-style API could allow sites to receive access to specific fonts that they need (like Consolas in the previous example) without exposing the list of installed fonts.
> 
> @plinss remembers a TPAC conversation from Fukuoka, perhaps in the Houdini TF, where a picker-style API got significant support.

We still believe that a picker-style API solves the most pressing user needs while having far preferable characteristics re: privacy, which we've documented in our [Web Platform Design Principles](https://w3ctag.github.io/design-principles/) document (specifically in [8.2. Use care when exposing APIs for selecting or enumerating devices](https://w3ctag.github.io/design-principles/#device-enumeration)).

Note that such an approach would also address WICG/local-font-access#36 and WICG/local-font-access#62.

From [this comment](https://github.com/WICG/local-font-access/issues/62#issuecomment-759620208) it looks like you're considering a picker-style API alternative. We look forward to learning more about it.

**Comment by oyiptong** [comment]
Indeed, we've been making progress with the API and have been prototyping a chooser behind a flag.
We're planning to launch another Origin Trial soon, with a modified API shape.

I'll be amending the explainer and spec copy soon.

**Comment by hober** [comment]
> Indeed, we've been making progress with the API and have been prototyping a chooser behind a flag.
> We're planning to launch another Origin Trial soon, with a modified API shape.
> 
> I'll be amending the explainer and spec copy soon.

Any updates for us?

**Comment by hober** [comment]
@plinss and I took another look at this today in our F2F and it seems we're still waiting on updates from @oyiptong. We're going to close this; please let us know when you're ready for us to take another look and we'll reopen it.

**Comment by inexorabletash** [comment]
@hober @torgo and anyone else - I commented over in #400 (sorry about the two issues!) but we're ready for another look. Pinging here hoping for attention too.
